### DISCUSSION TITLE
Collision Layers + Complex Queries

### DISCUSSION THREAD

**Discussion by BobbyAnguelov** [jrouwe/JoltPhysics#477] [General]
Hey, so I've been stuck on trying to figure out how to best implement my needs for physics queries and general collision given your layer system.  I'm gonna provide some examples of use cases below and would be curious what your solution for these problems in jolt would be and how you'd go about implementing it given the current APIs. Also sorry about the programmer art.

**First Example: Navigation**

![image](https://user-images.githubusercontent.com/10767490/226146994-165bc8da-3fb0-46bd-95fd-9fec19643e2d.png)

I need two distinct sets of collisions for player navigation. I need a 'simplified' collision (NAV) for general navigation/character controller queries and I need a more complex collision (ENV) more closely matching the visual mesh for IK (and in the case of shooting, bullet/visibility queries)

Some shapes would be flagged as ENV only, some would be flagged as NAV only and some shapes would be both ENV and NAV. From what I can understand from the code and the examples, one solution would be to create three narrow phase layers:

ENV,
NAV,
ENV+NAV

Then when I perform queries, I would use the NAV and ENV+NAV layers in the filter for navigation queries. For IK, I'd filter using ENV and ENV+NAV. 

So let's add a bit more complexity, we want another filter for the camera since some shapes need to block the character but not the camera, notice how the square shape at the top now doesnt block the camera.

![image](https://user-images.githubusercontent.com/10767490/226147193-32abad71-49d5-42ac-95bb-f25b56427c1a.png)

And to make things even more complex, when we are not blocked by the shape, we still need to detect that we are overlapping/touching any shapes flagged as ENV as we want to tell the engine to either hide or make those shapes transparent so as to not block the player view (think 3rd person game in a lot of interior environments). So we need the concept of a blocking collision and a non-blocking collision.

Would I just have more permutations of the layers, e.g.,

ENV, NAV, CAM, ENV+NAV, ENV+CAM, NAV+CAM, ENV+NAV+CAM

How would I specify the "overlap" case? Is that treating everything as colliding, and then filtering through the query/cast results and sorting by distance and manually flagging hits as blocking or non-blocking on the client side? How do we optimize this in terms of early-outing of queries when we hit a user defined blocking hit? I was looking at the CollisionCollector and the early out fraction stuff which seems like it's intended to be used for this assuming that all hits are added in sorted near to far order. Is this guaranteed?

**Second Example: Bullet Penetration and Visibility**

![image](https://user-images.githubusercontent.com/10767490/226147586-ff803f56-f59d-4fb1-bbb5-5742ecaa07da.png)

Imagine we have the scenario above, where we have multiple static ENV objects each with different query properties, some of the objects can optionally block either bullets or visibility. This kinda ties into the above scenario with the camera but there's a small difference:

For Visibility queries, it's a simple, cast a shape until the first blocking hit, we dont care about any overlaps. 
Whereas for bullets, we still want to terminate on the first blocking hit but we also want every single non-blocking hit (including back faces) along the way. 

Adding visibility and bullets to the layers explodes the permutations even more. And this can be a nightmare for artists to define the exact set of flags per object (ENV, NAV, CAM, VISIBILITY_BLOCKER, etc). How would you suggest handling this?

I'm currently thinking about exposing a set of flags, one per category (ENV, NAV, etc...) and having them tick the applicable ones. Then behind the scenes I will autogenerate all the layer permutations for the available set of flags and use those for jolt and these would never be exposed to the users or the programmers. Programmers would also go through a bitfield to layer resolution step when querying the physics scene. This doesnt feel super elegant to be honest and kinda feels like I'm hacking around what should be in my opinion very basic use cases for a game physics system. I feel like I'm missing something here?

Maybe I'm too used to unreal but there it's relatively simple as you can define the collision response per object, in terms of what it blocks (since under the hood there are no layers but rather bitmasks that are compared to determine collisions). I feel like I'm trying to force that approach on to jolt and it doesnt fit cleanly, so would love to know what the recommended best practice for these use cases is?

Any help would be much appreciated! Thanks!

**Comment by jrouwe** [comment]
> Would I just have more permutations of the layers, e.g., ENV, NAV, CAM, ENV+NAV, ENV+CAM, NAV+CAM, ENV+NAV+CAM

This would be one way of doing it and is roughly how we've done it, but we try to be smart about the amount of permutations we allow to not get an explosion of layers. E.g. you could have these layers:
* STATIC (your NAV)
* MOVEMENT_HELPER
* BULLET_BLOCKER (your ENV)
* CAMERA_BLOCKER
And have the following filters:
* Camera collision checks against STATIC, BULLET_BLOCKER and CAMERA_BLOCKER
* Bullets check against STATIC and BULLET_BLOCKER
* Character checks against STATIC and MOVEMENT_HELPER (and we would make the MOVEMENT_HELPER touch the top of each stair step instead of going through the center) 
* IK tests are e.g. against STATIC and BULLET_BLOCKER

You could also treat ObjectLayer as a bit mask and reserve 1 bit for NAV, 1 for ENV and 1 for CAM just like in the unreal solution you mention (there's nothing preventing you from having 65535 layers as long as you can define a sensible mapping to the broadphase layers). You could also use the Body and or Shape user data flags to do custom filtering and use the BodyFilter / ShapeFilter classes if you find the the ObjectLayer doesn't provide enough bits, in this case make sure the most important layers go to the ObjectLayer to filter out as much as possible as early as possible.

> How would I specify the "overlap" case? Is that treating everything as colliding, and then filtering through the query/cast results and sorting by distance and manually flagging hits as blocking or non-blocking on the client side?

Yes if you want to detect overlaps, you'll need to treat it as colliding and then ignore the hits in some parts of your code. I have to say that we unusually solve this by doing multiple queries with different filters as the systems responsible for these different behaviors are usually decoupled and do slightly different queries anyway.

> How do we optimize this in terms of early-outing of queries when we hit a user defined blocking hit? I was looking at the CollisionCollector and the early out fraction stuff which seems like it's intended to be used for this assuming that all hits are added in sorted near to far order. Is this guaranteed?

Jolt will add hits in a roughly sorted way (it sorts by bounding box), but it will definitely not report any hits further away than the early out fraction. If you want to do multiple queries in 1 query you will need to keep track of the furthest hit you're interested in yourself and update the early out fraction and you will continue to receive closer hits.

**Reply by BobbyAnguelov** [reply]
> This would be one way of doing it and is roughly how we've done it, but we try to be smart about the amount of permutations we allow to not get an explosion of layers. E.g. you could have these layers: STATIC (your NAV), MOVEMENT_HELPER, BULLET_BLOCKER (your ENV), CAMERA_BLOCKER

This is the part I don't understand and maybe I didn't explain properly. I can't have just those layers since I have objects that are in two layers at once: e.g., both a BULLET_BLOCKER and CAMERA_BLOCKER or both a STATIC and MOVEMENT_HELPER. This is where I'm struggling.  Just having the filter match multiple layers doesn't feel like enough granularity. Were all your physics objects exclusively in a single layer - did you never had objects that had multiple collision semantics? We had plenty of headaches in the past where we needed to disable camera collision on certain props sticking out of walls and so on. Even in your example, there's a problem, character cannot just check against STATIC and MOVEMENT_HELPER, since they will now go through all BULLET_BLOCKERs, since from what I understand bullet blockers are not STATIC or MOVEMENT_HELPERS. Once destructibles come into play this is further highlighted as some destructibles need to be BULLET_BLOCKERs but allow the player to pass through and trigger the destruction while other destructibles will not allow the player to pass through. 

Unreal has the distinction that they separate object type (your layer) and the collision response (filter bits) on a per object level. This makes it easy to handle edge/special case setups, as the artist simply says this is a static prop and it blocks X/Y/Z. This is very unreal specific but all the other places I worked it was with physx and so it was bitflags for the filters so each object could be part of multiple layers. I guess I can handle this in the BodyFilter? Doing this in the body filter concerns me since if I use the UserData as a ptr to additional information about the body, it's gonna pontentially be cache miss central per collision test as I'm now dereferencing that user data to get back some flags. Alternatively, I could use 32 bits of the user data for the flags and 32 bits as an index for actual user data which I allocate myself but that's a lot of extra complexity so it really feels like I will just abuse the "layers" to bit flags instead of explicit layers.

> Jolt will add hits in a roughly sorted way (it sorts by bounding box), but it will definitely not report any hits further away than the early out fraction. If you want to do multiple queries in 1 query you will need to keep track of the furthest hit you're interested in yourself and update the early out fraction and you will continue to receive closer hits.

So from what I understand, as we add hits to the collector, I will track the closest "blocking" hit internally and update the fraction based on it and this act as an additional "broad" rejection for further hits without entering any of the fine grain filtering code? So this is a not avery common use case since most of the time, I am already settings the maximum range for the query via the distance for the cast? So the fraction is something that most people can just ignore 90% of the time?

Sorry for the potentially dumb questions but I've not really encountering this approach to filtering before (mostly have experience with Havok and PhysX) and I'm not exactly a physics programmer either (more like a semi-power-user of the APIs). This also feels like another topic where we could expand on the examples or the documentation to make this easier for other users. Unreal has a pretty good writeup for this topic: [Unreal Collision Filtering](https://www.unrealengine.com/en-US/blog/collision-filtering), it would be nice to do the same. I dont mind writing that up once I understand the system and the best practices.

**Reply by jrouwe** [reply]
> Were all your physics objects exclusively in a single layer - did you never had objects that had multiple collision semantics?

Yes we have that, that's why we have some combination layers too.

But it sounds like you really want the Unreal way of doing things and I think it's perfectly fine to use ObjectLayer as a bitfield to store the 'Trace Channel' (if we stick with Unreal terminology), so you can have 16 bits in there and you can use the ObjectLayerFilter to determine if something collides or not (this has a virtual function so under your control).

Then we have the 'Object Channels' left, you can either place them in the User Data of a Body or perhaps the Shape of the body. Pointers are usually 16 byte aligned, so you can at least have 16 'Object Channels' in there without losing the ability to store a pointer. But if you want something that is inside Body you can also use Body::mCollisionGroup which holds 64 bits of user data. GroupFilters do during simulation what BodyFilters do during collision queries, but those bits are free to use, so you could reserve some bits there (that does mean you can't use the default filter GroupFilterTable anymore as that makes assumptions on what the bits are for).

Then the distinction between overlapping and colliding: 

* For rigid body simulation, ContactSettings::mIsSensor can turn a contact into a just a collision callback without a collision response. 
* If this is needed for CharacterVirtual too, then I can easily add the same bool to CharacterContactSettings.
* For collision queries, you're fully in control yourself as you can do whatever you want in CollisionCollector::AddHit.

> So the fraction is something that most people can just ignore 90% of the time?

Let's say you shoot a bullet over a distance of 100m, if it first collides with a bush after 10m then we don't need to update the early out fraction, we can just record an 'overlap'. Let's say it then finds another collision with a solid wall at 20m for which you would record a 'hit' in your AddHit function. If you then don't update the early out fraction it may continue to hit another wall at 30, 40, ..., 100m, so I think there is a clear benefit in properly managing the early out fraction. Note that since the sorting is not 100% accurate, you may get another AddHit callback with another solid wall at 15m after the hit at 20m, so in this case you would need to overwrite the 'hit' that you stored earlier. Also there may have been a bush at 17m that you stored an 'overlap' for that now also needs to be discarded because it is behind the 'hit'. The default collision collectors don't make the distinction between 'overlap' and 'hit', but you can easily create your own CollisionCollector that does it.

**Reply by BobbyAnguelov** [reply]
Okay yeh, the fraction thing is a brainfart on my side, it's obviously need in most cases!

Looking at the definition of the Body class, there already is enough space to increase the objectlayer to 32bits instead of 16bit and then it would be trivial to take the approach that Erin is suggesting and that I feel makes sense for me too.

![image](https://user-images.githubusercontent.com/10767490/226213570-8d249ed9-5905-44bd-871a-09ddb2e01d8f.png)

You're already padding out the struct in the double and 32bit scenarios and in the default one the space is already there.

Would it be possible to make that change? Or are there any knockon effects for that? Trying to reuse user data or the misuse the group filter for this seems like a less elegant and potentially less performant solution (depending on what you do with the userdata)

**Reply by jrouwe** [reply]
It's possible to change it and indeed it does not change the size of Body, but it does change the size of QuadTree::Tracking from 8 to 12 bytes (it was chosen to fit in the padding). On startup, an array of size MaxBodies of these Tracking structs is allocated, so in our case this would mean that we would allocate 480 KB extra, which I'm hesitant to just throw away (it's not only the allocation, but it will come with extra cache misses as well when working with the broadphase).

The only thing that needs to change to switch to 32 bits is [this](https://github.com/jrouwe/JoltPhysics/blob/352550fdd562ce6034b27ab12d1470840e879b56/Jolt/Physics/Collision/ObjectLayer.h#L12) and [this](https://github.com/jrouwe/JoltPhysics/blob/352550fdd562ce6034b27ab12d1470840e879b56/Jolt/Physics/Collision/ObjectLayer.h#L15) and the 2nd line should really have been defined as ~ObjectLayer(0) (I'll make that change).

Are you getting the library through CMake or vcpck or some other packaging tool that takes it directly from github? If so, I could make a define for this, if not then please just change that line and you pay for the extra memory but you get what you want.

**Reply by BobbyAnguelov** [reply]
Having that as a define would be great since I am using the cmake version and I am already building it locally since I needed to disable the whole program optimizing for the release build.

**Reply by jrouwe** [reply]
Please update to 364071164528f173539d49c7043e94798a40473d

**Comment by erincatto** [comment]
I've generally used category bits and mask bits to solve many of the problems Bobby has. So each object has some category bits and some mask bits. Then object-object collision culling is handled by the simple rule:
```
bool collide = (objectA.categoryBits & objectB.maskBits) != 0 && (objectB.categoryBits & objectA.maskBits) != 0;
```
This requires no collision matrix like layers do. The downside is it requires more bits. But not much more. Overwatch shipped with 16bits each for category and mask bits. A negligible amount of memory compared to everything else in a game these days. 
The key is to use categories that are unencumbered. For example, almost all static collision is just `STATIC`. Camera, projectile, nav, etc are all handled by the mask bits. For example:
low_res_collision:
```
categoryBits = STATIC
maskBits = CAMERA | PLAYER | DYNAMIC | DEBRIS
```
high_res_collision:
```
categoryBits  = STATIC
maskBits = PROJECTILE
```
invisible_player_blocker:
```
categoryBits = STATIC
maskBits = PLAYER
```
In Jolt if you allowed ObjectLayer to be more than 16 bits, say 32 bits or 64 bits then someone easily choose to use the category/mask approach. 8 bits each for category/mask is insufficient for most games. I think this could also be implemented fairly efficiently and avoid expensive callbacks.
Also notice the names above: `low_res_collision`, `high_res_collision`, `invisible_player_blocker`. These are basically _collision presets_ that can be managed on the game engine side with no awareness by the physics engine. These create reusable bit patterns that can be assigned to objects in the game editor. Thus these can automatically pick up changes to the category/mask bits without having to patch lots of assets.

**Reply by jrouwe** [reply]
Do I understand correctly that PROJECTILE would be a combination of bits from the categoryBits, so PROJECTILE = STATIC | SOME_OTHER_BIT? And that each category is a single bit?

We used to have the query bits in the objects too, but we realized that when doing a query, the flags of the object you query with (your maskBits) can be stored in the filter, so you only really need to store the bits of the objects you're querying against (your categoryBits). In case of a collision between two bodies in the simulation, the objects A and B can be arbitrarily swapped so you do need to store more information which is why there are more bits in the CollisionGroups.

**Reply by erincatto** [reply]
Queries had the same filtering. For example _I am a PROJECTILE_ and _I want to collide with DEBRIS and STATIC_. Here is my recollection of the set of category bits we used.
```
enum Categories
{
    STATIC = 0x0001,
    TEAM1 = 0x0002,
    TEAM2 = 0x0004,
    DYNAMIC = 0x0008,
    PROJECTILE = 0x0010,
    LINE_OF_SIGHT = 0x0020,
    DEBRIS = 0x0040,
    TURRET = 0x0080
    BREAKABLE = 0x0100
    WATER = 0x0200
}
```
So query-versus-object uses:
```
bool collide = (query.categoryBits & object.maskBits) != 0 && (object.categoryBits & query.maskBits) != 0;
```
Queries often have more complex filtering rules than object-vs-object, and those are handled in the query (e.g. raycast) callbacks. Such as check against ignored entities. But the query filtering is still useful to reduce the number of callbacks.

**Reply by BobbyAnguelov** [reply]
This is what I'm used to as well, using the bits as a non-virtual pre-filter and then leaving the more complex filtering to a callback with a virtual function.

**Reply by jrouwe** [reply]
Thanks @erincatto for explaining!
